### ISSUE TITLE
Rule for `float(x)`, or `convert(AbstractArray, x)`

### ISSUE THREAD

**Issue by mcabbott** [JuliaDiff/ChainRules.jl#581]
```julia
julia> Zygote.gradient(x -> sum(abs2, float(x)), [1,2,3])
ERROR: Mutating arrays is not supported -- called copyto!(::Vector{Float64}, _...)
Stacktrace:
...
  [7] (::typeof(∂(AbstractVector{Float64})))(Δ::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/umM0L/src/compiler/interface2.jl:0
...
 [11] (::typeof(∂(convert)))(Δ::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/umM0L/src/compiler/interface2.jl:0
```
What this calls is `convert(AbstractArray{typeof(float(zero(T)))}, A)`, and probably that ought to have a rule, which would also fix these:
```julia
julia> gradient(x -> sum(abs2, convert(Array{Float32}, x)), [1,2,3])
([2.0, 4.0, 6.0],)

julia> gradient(x -> sum(abs2, convert(AbstractArray{Float32}, x)), [1,2,3])
ERROR: Mutating arrays is not supported -- called copyto!(::Vector{Float32}, _...)

julia> gradient(x -> sum(abs2, Array{Float32}(x)), [1,2,3])
([2.0, 4.0, 6.0],)

julia> gradient(x -> sum(abs2, AbstractArray{Float32}(x)), [1,2,3])
ERROR: Mutating arrays is not supported -- called copyto!(::Vector{Float32}, _...)
```

**Comment by mzgubic** [comment]
why has this been closed?

**Comment by mcabbott** [comment]
Sorry, should have commented, fixed in #565 but didn't make the link.
